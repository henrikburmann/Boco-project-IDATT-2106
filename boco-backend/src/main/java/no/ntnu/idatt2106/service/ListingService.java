package no.ntnu.idatt2106.service;

import no.ntnu.idatt2106.model.DAO.CommunityDAO;
import no.ntnu.idatt2106.model.DAO.CommunityListingDAO;
import no.ntnu.idatt2106.model.DAO.ListingDAO;
import no.ntnu.idatt2106.model.DAO.UserDAO;
import no.ntnu.idatt2106.model.DTO.ListingDTO;
import no.ntnu.idatt2106.repository.ListingRepository;


import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Service;

/**
 * This class is uses the access-point to the listing table in the DB.
 * This class uses the methods from {@link no.ntnu.idatt2106.repository.ListingRepository ListingRepository}
 */
@Service
public class ListingService {
    private final ListingRepository listingRepository;
    private final CommunityListingService communityListingService;
    private final ListingCategoryService listingCategoryService;

    public ListingService(ListingRepository listingRepository, CommunityListingService communityListingService,
                          ListingCategoryService listingCategoryService) {
        this.listingRepository = listingRepository;
        this.communityListingService = communityListingService;
        this.listingCategoryService = listingCategoryService;
    }

    /**
     * Saves a ListingDAO to the DB
     * @param listingDAO
     */
    public void saveListing(ListingDAO listingDAO) {
        listingRepository.save(listingDAO);
    }

    /**
     * Finds all Listings in the Listing table
     * @return All listings
     */
    public List<ListingDAO> getAllListings() {
        return listingRepository.findAllByDeletedIsFalse();
    }

    /**
     * Finds all listing posted by a user
     * @param userDAO
     * @return All the user's listings
     */
    public List<ListingDAO> getAllOfUsersListings(UserDAO userDAO) {
        return listingRepository.findAllFromListingDAOByUser(userDAO);
    }

    /**
     * Retrieves all listings where variable deleted is false.
     * @param userDAO
     * @return
     */
    public List<ListingDAO> getAllOfNonDeletedListings(UserDAO userDAO) {
        //Finds all user listings
        List<ListingDAO> allUserListings = listingRepository.findAllFromListingDAOByUser(userDAO);
        List<ListingDAO> allNonDeletedUserListings = new ArrayList<>();
        //Finds all user listings where deleted is false
        for (ListingDAO listing : allUserListings) {
            if (!listing.isDeleted()) {
                allNonDeletedUserListings.add(listing);
            }
        }
        return allNonDeletedUserListings;
    }

    /**
     * Finds a specific listing
     * @param listingID
     * @return An optional with a specific listing
     */
    public ListingDAO getListingDAOByID(int listingID) {
        return listingRepository.findListingDAOByListingID(listingID);
    }

    /**
     * Converts a list of ListingDAOs to ListingDTOs using the
     * convertOneListingDAOToDTO for every Object
     * @param listingDAOs The list of DAOs that is to be converted
     * @return A list of converted DTOs.
     */
    public List<ListingDTO> convertMultipleFromListingDAOToDTO(List<ListingDAO> listingDAOs) {
        List<ListingDTO> listingDTOs = new ArrayList<>();
        for (ListingDAO listingDAO : listingDAOs) {
            listingDTOs.add(convertOneListingDAOToDTO(listingDAO));
        }
        return listingDTOs;
    }

    /**
     * Method for converting a ListingDAO to a ListingDTO.
     * @param listingDAO The ListingDAO that is to be converted
     * @return The converted ListingDAO, now a DTO
     */
    public ListingDTO convertOneListingDAOToDTO(ListingDAO listingDAO) {
        // Finds all the listing's categorynames through the listingCategory junction
        // table
        String[] categoryNames = listingCategoryService.getCategoryNamesByListingID(listingDAO);
        // Finds all the listing's communityIDs through the communityListing junction
        // table
        int[] communityIDs = communityListingService.getCommunityListingIDsByListingID(listingDAO);
        // Creates a ListingDTO with the DAO-information, aswell as the categorynames
        // and communityIDs.
        ListingDTO listingDTO = new ListingDTO(listingDAO.getTitle(), listingDAO.getDescription(),
                listingDAO.getPricePerDay(), listingDAO.getAddress(), listingDAO.getUser().getUserID(), categoryNames,
                communityIDs);
        listingDTO.setListingID(listingDAO.getListingID());
        return listingDTO;
    }

    /**
     * A method for finding a single listing instance with a given listing id.
     * @param listingId The id of the listing to search for.
     * @return Returns a listing dao with id matching the given id.
     */
    public ListingDAO findListingByListingId(int listingId) {
        return listingRepository.findListingDAOByListingID(listingId);
    }

    /**
     * A method for finding all listing instances for a given user DAO.
     * Returns a list of all listing daos in the db with this owner.
     * @param ownerId The UserDAO for the owner of the items.
     * @return Returns a list of all listing daos in the db with this owner dao.
     */
    public List<ListingDAO> findAllListingDAOByIdOfOwner(UserDAO ownerId) {
        return listingRepository.findAllByUserAndDeletedIsFalse(ownerId);
    }

    /**
     * Find all listings posted by a user
     * @param user user to find listings for
     * @return A list of all the users ListingDAOs
     */
    public List<ListingDAO> findListingsByUserDAO(UserDAO user) {
        return listingRepository.findListingDAOSByUser(user);
    }

    /**
     * Returns a list of ListingDTOs with title containing requested phrase.
     * @param title
     * @return
     */
    public List<ListingDTO> getListingDTOByTitle(String title) {
        //Gets all lisitngDAOs with the requested title
        List<ListingDAO> listingDAOs = listingRepository.findAllByTitleLike(title);
        //Converts all the DAOs to DTOs
        List<ListingDTO> listingDTOs =
                convertMultipleFromListingDAOToDTO(listingDAOs);
        return listingDTOs;
    }

    /**
     * Get all listings from a community
     * @param communityDAO Community to find listings for
     * @return List of ListingDAOs from the community
     */
    public List<ListingDAO> getAllListingsInACommunity(CommunityDAO communityDAO) {
        List<ListingDAO> listings = new ArrayList<>();
        List<CommunityListingDAO> communityListings = communityListingService
                .getAllCommunityListingForCommunity(communityDAO);

        if (communityListings != null) {
            for (int i = 0; i < communityListings.size(); i++) {
                listings.add(communityListings.get(i).getListing());
            }
        }
        return listings;
    }

    /**
     * Convert a list of ListingDAOs to a list of ListingDTOs
     * @param list List of ListingDAOs to convert
     * @return A list of ListingDTOs
     */
    public List<ListingDTO> convertListOfListingDAOToListOfListingDTO(List<ListingDAO> list) {
        List<ListingDTO> convertedList = new ArrayList<>();
        for (int i = 0; i < list.size(); i++) {
            convertedList.add(new ListingDTO(list.get(i)));
        }
        return convertedList;
    }

    /**
     * Gets the previous listing posted by a user
     * @param userID user to find listing for
     * @return ListingDAO that is the users last posted listing
     */
    public ListingDAO getUsersLastPostedListing(int userID) {
        return listingRepository.findLastAddedListingByUser(userID);
    }

    /**
     * Sets all listings made by a user to isDeleted = true
     * @param userDAO user to delete listings for
     */
    public void deleteListingsForUser(UserDAO userDAO) {
        List<ListingDAO> listings = getAllOfUsersListings(userDAO);
        for (ListingDAO listing : listings) {
            listing.setDeleted(true);
        }
    }
}
