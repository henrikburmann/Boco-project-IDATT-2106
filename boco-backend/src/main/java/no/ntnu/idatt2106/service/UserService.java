package no.ntnu.idatt2106.service;

import no.ntnu.idatt2106.model.DAO.UserDAO;
import no.ntnu.idatt2106.model.DTO.UserDTO;
import no.ntnu.idatt2106.repository.UserRepository;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Base64;


import no.ntnu.idatt2106.util.HashUtil;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    /**
     * A method to find a specific user by their unique user id.
     * @param userId The unique user id of the user you want to find
     * @return Returns a UserDAO of the user matching the input user id.
     */
    public UserDAO findUserByUserId(int userId) {
        return userRepository.findUserDAOByUserID(userId);
    }

    /**
     * A method to find a user by their email.
     * @param email The email of the user you want to find
     * @return Returns a UserDAO for the user with this email.
     */
    public UserDAO findUserByEmail(String email) {
        return userRepository.findUserDAOByEmail(email);
    }

    /**
     * A method to save a new user to the DB.
     * @param user The UserDAO of the new user you want to store in the DB.
     */
    public void saveUser(UserDAO user) {
        userRepository.save(user);
    }

    /**
     * Converts a list of user DAOs to a list of user DTOs
     * @param list A list of user DAOs
     * @return A list of user DTOs
     */
    public List<UserDTO> convertListUserDAOToListUserDTO(List<UserDAO> list) {
        List<UserDTO> convertedList = new ArrayList<>();
        for (int i = 0; i < list.size(); i++) {
            convertedList.add(new UserDTO(list.get(i)));
        }
        return convertedList;
    }

    /**
     * Change a password for a user
     * @param userDAO  User to change password for
     * @param password Password to change to
     * @return UserDAO of the user that has gotten password changed
     */
    public UserDAO changePasswordForUser(UserDAO userDAO, String password) {
        HashUtil hashUtil = new HashUtil();
        byte[] salt = hashUtil.getRandomSalt();
        byte[] hashedPassword = hashUtil.getHashedPassword(password, salt);

        userDAO.setSalt(Base64.getEncoder().encodeToString(salt));
        userDAO.setHash(Base64.getEncoder().encodeToString(hashedPassword));

        saveUser(userDAO);
        return userDAO;
    }

    /**
     * Method that authenticates password
     * @param user     User that has the password
     * @param password Password to authenticate
     * @return boolean that explains if the operation succeeded or not
     */
    public boolean attemptAuthenticationOfPassword(UserDAO user, String password) throws NoSuchAlgorithmException {
        if (user == null) {
            return false;
        }
        MessageDigest md = MessageDigest.getInstance("SHA-512");
        md.update(Base64.getDecoder().decode(user.getSalt()));
        byte[] hashedPassword = Base64.getEncoder().encode(md.digest(password.getBytes(StandardCharsets.UTF_8)));

        if (new String(hashedPassword).equals(new String(user.getHash().getBytes(StandardCharsets.UTF_8)))) {
            return true;
        }
        return false;
    }

    /**
     * Method that clears all personal data from a user, this method is used when deleting a user to keep ratings given
     * @param userDAO User to clear
     */
    public void clearUserInfo(UserDAO userDAO) {
        userDAO.setFirstName("Slettet");
        userDAO.setLastName("Konto: " + userDAO.getUserID());
        userDAO.setPicture("");
        userDAO.setHash("");
        userDAO.setSalt("");
        userDAO.setAddress("");
        userDAO.setEmail("");
        userRepository.save(userDAO);
    }
}
