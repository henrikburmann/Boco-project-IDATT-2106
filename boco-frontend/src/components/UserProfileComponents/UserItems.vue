<template>
  <!-- Shows all the items a user has posted with search and pagination.
       Includes a dropdown menu for editing or deleting an item. -->
  <div id="headline" class="text-xl md:text-2xl text-primary-light font-medium">
    Mine gjenstander
  </div>
  <!-- Search field -->
  <div class="relative mx-4" id="searchComponent">
    <span class="absolute inset-y-0 left-0 flex items-center pl-3">
      <svg class="w-5 h-5 text-gray-400" viewBox="0 0 24 24" fill="none">
        <path
          d="M21 21L15 15M17 10C17 13.866 13.866 17 10 17C6.13401 17 3 13.866 3 10C3 6.13401 6.13401 3 10 3C13.866 3 17 6.13401 17 10Z"
          stroke="currentColor"
          stroke-width="2"
          stroke-linecap="round"
          stroke-linejoin="round"
        ></path>
      </svg>
    </span>

    <input
      type="text"
      id="searchInput"
      class="w-full py-3 pl-10 pr-4 text-gray-700 bg-white border rounded-md dark:bg-gray-800 dark:text-gray-300 dark:border-gray-600 focus:border-primary-medium dark:focus:border-primary-medium focus:outline-none focus:ring"
      placeholder="Søk"
      v-model="search"
      @change="searchWritten"
    />
  </div>
  <div class="absolute inset-x-0">
    <!-- ItemCards -->
    <div class="flex items-center justify-center w-screen">
      <!-- Shows items based on pagination -->
      <div
        class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full"
        v-if="showItems"
      >
        <div
          class="cardContainer"
          id="item"
          v-for="item in visibleItems"
          :key="item"
        >
          <div class="w-full">
            <ItemCard
              id="ItemCardPage"
              class="ItemCard w-full h-full"
              :item="item"
            />
          </div>

          <!-- Dropdown menu with options for editing an item and deleting an item -->
          <TripleDotButton class="DotButton" @click="openDotMenu(item)" />

          <div
            v-show="item.toggle"
            class="options z-10 w-44 text-base list-none bg-white rounded divide-y divide-gray-100 shadow dark:bg-gray-700 pt-4 pl-12"
          >
            <ul
              class="py-1 absolute bg-white ring-1 ring-gray-300 rounded-xl"
              aria-labelledby="dropdownDefault"
            >
              <li>
                <button
                  @click="
                    this.$router.push('/item/' + item.listingID + '/edit')
                  "
                  class="editButton block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  Rediger gjenstand
                </button>
              </li>
              <li>
                <button
                  @click="goToDeleteItem(item.listingID)"
                  class="deleteButton block py-2 px-4 text-sm text-error-medium hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  Slett gjenstand
                </button>
              </li>
            </ul>
          </div>
        </div>

        <!-- A waring asking the user if it is sure it wants to delete the item
             with options to go ahead with the deleting or to cancel the delete. -->
        <CustomFooterModal
          @close="this.readyToDelete = false"
          :visible="readyToDelete"
          :title="'Sikker på at du vil slette annonsen?'"
          :message="''"
        >
          <div class="flex justify-center p-2">
            <ColoredButton
              id="#cancelDeleteButton1"
              :text="'Avbryt'"
              @click="cancelDelete"
              class="bg-gray-500 m-2"
            ></ColoredButton>

            <ColoredButton
              id="confirmDeleteButton1"
              @click="deleteItem"
              :text="'Slett'"
              class="confirmDelete m-2 bg-error-medium"
            >
            </ColoredButton>
          </div>
        </CustomFooterModal>
      </div>

      <!-- Shows items based on search field input -->
      <div
        class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full"
        v-if="showSearchedItems"
      >
        <div
          class="cardContainer"
          id="item"
          v-for="item in searchedItems"
          :key="item"
        >
          <div class="w-full">
            <ItemCard
              id="ItemCardSearch"
              class="ItemCard w-full h-full"
              :item="item"
            />
          </div>

          <!-- Dropdown menu with options for editing an item and deleting an item -->
          <TripleDotButton class="DotButton" @click="openDotMenu(item)" />
          <div
            v-show="item.toggle"
            class="options z-10 w-44 text-base list-none bg-white rounded divide-y divide-gray-100 shadow dark:bg-gray-700"
          >
            <ul
              class="py-1 absolute bg-white ring-1 ring-gray-300 rounded-xl"
              aria-labelledby="dropdownDefault"
            >
              <li>
                <button
                  @click="
                    this.$router.push('/item/' + item.listingID + '/edit')
                  "
                  class="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  Rediger gjenstand
                </button>
              </li>
              <li>
                <button
                  @click="goToDeleteItem(item.listingID)"
                  class="block py-2 px-4 text-sm text-error-medium hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  Slett gjenstand
                </button>
              </li>
            </ul>
          </div>
        </div>

        <!-- A waring asking the user if it is sure it wants to delete the item
             with options to go ahead with the deleting or to cancel the delete. -->
        <CustomFooterModal
          @close="this.readyToDelete = false"
          :visible="readyToDelete"
          :title="'Sikker på at du vil slette annonsen?'"
          :message="''"
        >
          <div class="flex justify-center p-2">
            <ColoredButton
              id="cancelDeleteButton"
              :text="'Avbryt'"
              @click="cancelDelete"
              class="bg-gray-500 m-2"
            ></ColoredButton>

            <ColoredButton
              id="confirmDeleteButton"
              @click="deleteItem"
              :text="'Slett'"
              class="m-2 bg-error-medium"
            >
            </ColoredButton>
          </div>
        </CustomFooterModal>
      </div>
    </div>

    <!-- pagination -->
    <div class="flex justify-center" v-if="showItems">
      <PaginationTemplate
        v-bind:items="items"
        v-on:page:update="updatePage"
        v-bind:currentPage="currentPage"
        v-bind:pageSize="pageSize"
        class="mt-10"
      />
    </div>
  </div>
</template>
<script>
import TripleDotButton from "@/components/BaseComponents/TripleDotButton.vue";
import ColoredButton from "@/components/BaseComponents/ColoredButton.vue";
import ItemCard from "@/components/ItemComponents/ItemCard.vue";
import PaginationTemplate from "@/components/BaseComponents/PaginationTemplate";
import CustomFooterModal from "@/components/BaseComponents/CustomFooterModal.vue";

import UserService from "@/services/user.service";
import listingService from "@/services/listing.service";

export default {
  name: "UserItems",
  components: {
    ItemCard,
    TripleDotButton,
    PaginationTemplate,
    CustomFooterModal,
    ColoredButton,
  },
  data() {
    return {
      items: [],
      item: {
        listingID: 0,
        img: "",
        address: "",
        title: "",
        pricePerDay: 0,
        toggle: false,
      },
      chosenItem: null,
      showItems: true,
      showSearchedItems: false,
      search: "",
      readyToDelete: false,
      dropdown: false,

      //Variables connected to pagination
      currentPage: 0,
      pageSize: 12,
      visibleItems: [],
    };
  },
  computed: {
    /**
     * Searchs items based on their title, address and price per day.
     */
    searchedItems() {
      let filteredItems = [];

      filteredItems = this.items.filter(
        (p) =>
          p.title.toLowerCase().includes(this.search.toLowerCase()) ||
          p.address.toLowerCase().includes(this.search.toLowerCase()) ||
          p.pricePerDay === Number(this.search)
      );

      return filteredItems;
    },
  },
  methods: {
    openDotMenu(item) {
      if (item.toggle == false) {
        for (var i = 0; i < this.visibleItems.length; i++) {
          this.visibleItems[i].toggle = false;
        }
        item.toggle = true;
      } else {
        item.toggle = false;
      }
    },
    getUserListingsFromAPI: async function () {
      this.items = await UserService.getUserListings();
      for (var i = 0; i < this.items.length; i++) {
        this.items[i].toggle = false;
        let images = await listingService.getItemPictures(
          this.items[i].listingID
        );
        if (images.length > 0) {
          this.items[i].img = images[0].picture;
        }
      }
    },
    cancelDelete() {
      this.readyToDelete = false;
    },
    /**
     * 2 methods related to pagination. Updates page and updates
     * visible items on page.
     */
    updatePage(pageNumber) {
      this.currentPage = pageNumber;
      this.updateVisibleTodos();
    },
    updateVisibleTodos() {
      this.visibleItems = this.items.slice(
        this.currentPage * this.pageSize,
        this.currentPage * this.pageSize + this.pageSize
      );

      // if we have 0 visible items, go back a page
      if (this.visibleItems.length === 0 && this.currentPage > 0) {
        this.updatePage(this.currentPage - 1);
      }
    },

    goToDeleteItem(item) {
      this.chosenItem = item;
      this.readyToDelete = true;
    },
    async deleteItem() {
      await UserService.setListingToDeleted(this.chosenItem);
      this.$router.go(0);
    },

    /**
     * This method triggers when search input field is changed
     */
    searchWritten: function () {
      if (this.search.length > 0) {
        this.showItems = false;
        this.showSearchedItems = true;
      } else {
        this.showItems = true;
        this.showSearchedItems = false;
      }
    },
  },

  /**
   * Gets userlistings and prepares the pagination by
   * updating the items to be visible.
   */
  async beforeMount() {
    await this.getUserListingsFromAPI();
    this.updateVisibleTodos();
  },
};
</script>

<style>
#headline {
  display: block;
  margin-top: 10px;
  margin-bottom: 10px;
  margin-left: 20px;
}
.cardContainer {
  position: relative;
}

.DotButton {
  position: absolute;
  right: 40px;
  bottom: 10px;
}
.options {
  position: absolute;
}
</style>
