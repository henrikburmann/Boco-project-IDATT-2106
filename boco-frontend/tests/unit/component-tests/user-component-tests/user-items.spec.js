import { mount } from "@vue/test-utils";
import UserItems from "@/components/UserProfileComponents/UserItems.vue";
import UserService from "@/services/user.service";
import ListingService from "@/services/listing.service";

jest.mock("@/services/user.service", () => jest.fn());
jest.mock("@/services/listing.service", () => jest.fn());

UserService.getUserListings = jest.fn(() => {
  return new Promise((resolve) => {
    resolve([
      {
        listingID: 1,
        img: "",
        address: "Veien",
        title: "Matboks",
        pricePerDay: 50,
        toggle: false,
      },
    ]);
  });
});

UserService.setListingToDeleted = jest.fn(() => {
  return new Promise((resolve) => {
    resolve();
  });
});

ListingService.getItemPictures = jest.fn(() => {
  return new Promise((resolve) => {
    resolve([]);
  });
});

jest.mock("axios");

let mockRouter;
let wrapper;

describe("UserItems component", () => {
  mockRouter = {
    push: jest.fn(),
    go: jest.fn(),
  };

  beforeEach(() => {
    wrapper = mount(UserItems, {
      global: {
        mocks: {
          $router: mockRouter,
        },
      },
      data() {
        return {
          showItems: true,
        };
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("is instantiated", () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  it("Check headline", () => {
    expect(wrapper.find("#headline").text()).toMatch("Mine gjenstander");
  });

  it("Check if pressing 'Rediger gjenstand' pushes to right path", async () => {
    const dotButton = wrapper.findAll(".DotButton")[0];
    dotButton.trigger("click");
    const editButton = wrapper.findAll(".editButton")[0];
    expect(editButton.text()).toBe("Rediger gjenstand");
    editButton.trigger("click");
    await wrapper.vm.$nextTick();
    expect(mockRouter.push).toBeCalled();
    expect(mockRouter.push).toBeCalledWith("/item/1/edit");
  });

  it("Check if delete is called", async () => {
    const dotButton = wrapper.findAll(".DotButton")[0];
    dotButton.trigger("click");
    await wrapper.vm.$nextTick();

    const deleteButton = wrapper.findAll(".deleteButton")[0];
    expect(deleteButton.text()).toBe("Slett gjenstand");
    deleteButton.trigger("click");
    await wrapper.vm.$nextTick();

    const confirmButton = wrapper.findAll(".confirmDelete")[0];
    expect(confirmButton.text()).toBe("Slett");

    //Feiler herfra
    confirmButton.trigger("click");
    await wrapper.vm.$nextTick();

    expect(ListingService.getItemPictures).toBeCalled();
    expect(mockRouter.go).toBeCalled();
  });
});
