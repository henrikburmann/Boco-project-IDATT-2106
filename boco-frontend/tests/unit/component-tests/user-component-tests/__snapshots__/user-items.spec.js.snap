// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserItems component renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <!-- Shows all the items a user has posted with search and pagination.
       Includes a dropdown menu for editing or deleting an item. -->
  <div
    class="text-xl md:text-2xl text-primary-light font-medium"
    id="headline"
  >
     Mine gjenstander 
  </div>
  <!-- Search field -->
  <div
    class="relative mx-4"
    id="searchComponent"
  >
    <span
      class="absolute inset-y-0 left-0 flex items-center pl-3"
    >
      <svg
        class="w-5 h-5 text-gray-400"
        fill="none"
        viewBox="0 0 24 24"
      >
        <path
          d="M21 21L15 15M17 10C17 13.866 13.866 17 10 17C6.13401 17 3 13.866 3 10C3 6.13401 6.13401 3 10 3C13.866 3 17 6.13401 17 10Z"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </span>
    <input
      class="w-full py-3 pl-10 pr-4 text-gray-700 bg-white border rounded-md dark:bg-gray-800 dark:text-gray-300 dark:border-gray-600 focus:border-primary-medium dark:focus:border-primary-medium focus:outline-none focus:ring"
      id="searchInput"
      placeholder="SÃ¸k"
      type="text"
    />
  </div>
  <div
    class="absolute inset-x-0"
  >
    <!-- ItemCards -->
    <div
      class="flex items-center justify-center w-screen"
    >
      <!-- Shows items based on pagination -->
      <div
        class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full"
      >
        
        <div
          class="cardContainer"
          id="item"
        >
          <div
            class="w-full"
          >
            
            <!-- Item card, displays title, address, price per day and a picture -->
            <div
              class="mt-5 px-5 ItemCard w-full h-full"
              id="ItemCardPage"
            >
              <div
                class="w-full h-full rounded bg-gray-100 ring-1 ring-gray-200 overflow-hidden display:inline-block correct-size cursor-pointer"
              >
                <div
                  class="relative h-0 pb-[66.7%]"
                >
                  <img
                    alt="Item image"
                    class="w-full h-full object-contain absolute"
                    src=""
                  />
                </div>
                <div
                  class="p-1 m-1 bottom-0"
                >
                  <p
                    class="font-bold text-sm"
                    id="title"
                  >
                    Matboks
                  </p>
                  <p
                    class="text-gray-700 text-xs"
                    id="price"
                  >
                    50 kr 
                  </p>
                  <p
                    class="text-gray-700 text-xs font-bold"
                    id="adress"
                  >
                    Veien
                  </p>
                </div>
              </div>
            </div>
            
          </div>
          <!-- Dropdown menu with options for editing an item and deleting an item -->
          <button
            class="w-10 h-10 text-gray-500 dark:text-gray-400 hover:bg-gray-100 dark:hover:bg-gray-700 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-700 rounded-lg float-right text-sm p-1.5 DotButton"
            data-dropdown-toggle="dropdown"
            id="dropdownDefault"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 5v.01M12 12v.01M12 19v.01M12 6a1 1 0 110-2 1 1 0 010 2zm0 7a1 1 0 110-2 1 1 0 010 2zm0 7a1 1 0 110-2 1 1 0 010 2z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="options z-10 w-44 text-base list-none bg-white rounded divide-y divide-gray-100 shadow dark:bg-gray-700 pt-4 pl-12"
            style="display: none;"
          >
            <ul
              aria-labelledby="dropdownDefault"
              class="py-1 absolute bg-white ring-1 ring-gray-300 rounded-xl"
            >
              <li>
                <button
                  class="editButton block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                   Rediger gjenstand 
                </button>
              </li>
              <li>
                <button
                  class="deleteButton block py-2 px-4 text-sm text-error-medium hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                   Slett gjenstand 
                </button>
              </li>
            </ul>
          </div>
        </div>
        
        <!-- A waring asking the user if it is sure it wants to delete the item
             with options to go ahead with the deleting or to cancel the delete. -->
        
        <!-- Main modal -->
        <!--v-if-->
        
      </div>
      <!-- Shows items based on search field input -->
      <!--v-if-->
    </div>
    <!-- pagination -->
    <div
      class="flex justify-center"
    >
      
      <!-- Pagination -->
      <div
        class="mt-10"
      >
        <!-- Prev button -->
        <!--v-if-->
        <!-- Current page -->
        <label
          class="mx-2 text-primary-light"
        >
          1 av 1
        </label>
        <!-- Next button -->
        <!--v-if-->
      </div>
      
    </div>
  </div>
  
</div>
`;
